rustls/tests/api.rs

```diff
@@ -479,7 +479,9 @@ fn server_allow_any_anonymous_or_authenticated_client() {
     let kt = KeyType::Rsa;
     for client_cert_chain in [None, Some(kt.get_client_chain())].iter() {
         let client_auth_roots = get_client_root_store(kt);
-        let client_auth = AllowAnyAnonymousOrAuthenticatedClient::new(client_auth_roots);
+        // TODO(@cpu): allow building test config with CRLs.
+        let client_auth =
+            AllowAnyAnonymousOrAuthenticatedClient::new(client_auth_roots, Vec::default()).unwrap();
```

Comment:
Looking ahead to test coverage, how would folks like to see test CRL generation handled?

I can think of a few options:
1. We follow the `webpki` lead and provide a Python script using `cryptography` to generate test CRLs the rust code can use and check those in.
2. We use `rcgen` to generate test CRLs with rust at runtime, taking a dev-dependency on `rcgen`.
3. We use `openssl` (maybe with a small helper script) to generate test CRLs and check those in.
4. ???

I think option 2 is the nicest, but requires a yet-to-be-merged or released branch of `rcgen` that might hold up this work :-/

Option 3 is probably my second favourite since there are already some test resources being generated this way.

Option 1 is crummy. We'd like to move away from this in `webpki` and so it doesn't seem forward looking to add more Python here.

Reply:
Assuming your rcgen branch will be merged eventually (which seems like a reasonable assumption, assuming you haven't gotten pushback on the idea of merging something like it) I'd be okay taking on a temporary dependency on a Git commit from your branch. Given this is just a dev-dep, I don't think it should hold up publishing to crates.io?

Reply:
I think the goals of this should be:

# tests are generated by an independent, quality implementation
# there's some chance the test data reflect how we expect people to use the feature
# the tests are as deterministic and fast as reasonably possible

I'm not sure rcgen (being backed by ring apart from anything else) would be independent or common enough?

Reply:
It's fair that our own implementation (even though reviewed by a separate maintainer) wouldn't be very independent, and I can see how that might be a reason to go with Python or OpenSSL instead. I wonder if the Rust openssl crate would expose support for this kind of stuff.

Reply:
I was able to use openssl-rs to parse CRLs in [an rcgen test](https://github.com/est31/rcgen/blob/3a8535a0724963e62dc3828ad8c73c957396a883/tests/openssl.rs#L391-L429) but I don't see any API surface for *generating* CRLs. I think I would probably have to use the `openssl` CLI (:nauseated_face:), or get into the weeds with new code for `openssl-sys` and `openssl-rs`.

I think I'd lightly prefer a small standalone `.py` file over bash+openssl if `rcgen` isn't independent enough. While tempting to get into the `openssl-rs` stuff I think I'd prefer not to block this PR on making new contributions to two codebases I haven't worked in before. That seems like better for a rainy day follow-up.

Reply:
I'd be okay with that.

Reply:
> I think I'd lightly prefer a small standalone .py file over bash+openssl if rcgen isn't independent enough.

Digging into the existing API tests more I've come around to updating the `build-a-pki.sh` script that's already tooled around `openssl`. I think it'll be a pretty minimal diff and will fit into the existing approach better than inventing something new and retrofitting.

Reply:
`openssl` never makes life easy, but I managed to get something workable going without too much fuss: https://github.com/rustls/rustls/pull/1326/commits/590c9c29fa1af1b15b83b2bc820e1756f839a4d5

Bonus fact: the default `openssl.cnf` configuration still [carries a workaround for Netscape Communicator](https://github.com/openssl/openssl/blob/8b7d5ea7dd602eb7c2c4bc5ad45489dc5fc711f6/apps/openssl.cnf#L107-L108) (last released ~Aug 2002) :roll_eyes: